cartridge/roles/expirationd.lua

```diff
@@ -0,0 +1,209 @@
+        if not ok then
+            error(res)
+        end
+        opts[opt] = res
+    end
+
+    return opts
+end
+
+local function get_task_config(task_conf)
+    local conf = table.deepcopy(task_conf)
+    local params_map = {
+        space_id = {required = true, types = {"n", "s"}},
+        is_expired = {required = true, types = {"f"}},
+        is_master_only = {required = false, types = {"b"}},
+        options = {required = false, types = {"t"}},
+    }
+    for k, _ in pairs(conf) do
+        if type(k) ~= "string" then
+            error(role_name .. ": param must be a string")
+        end
+        if params_map[k] == nil then
+            error(role_name .. ": unsupported param " .. k)
+        end
+    end
+
+    for param, opts in pairs(params_map) do
+        if opts.required and conf[param] == nil then
+            error(role_name .. ": " .. param .. " is required")
+        end
+        if conf[param] ~= nil then
+            local ok, res = get_param(param, conf[param], opts.types)
+            if not ok then
+                error(res)
+            end
+            conf[param] = res
+        end
+    end
+
+    conf.options = get_task_options(conf.options)
+    return conf
+end
+
+local function init()
+
+end
+
+local function validate_config(conf_new)
+    local conf = table.deepcopy(conf_new[role_name] or {})
+
+    for task_name, task_conf in pairs(conf) do
+        local ok, res = get_param("task name", task_name, {"s"})
+        if not ok then
+            error(res)
+        end
+        get_task_config(task_conf)
+    end
+
+    return true
+end
+
+local function apply_config(conf_new, opts)
+    local conf = table.deepcopy(conf_new[role_name] or {})
+
+    -- finishes tasks from an old configuration
+    for i=#started,1,-1 do
+        local task_name = started[i]
+        local ok, _ = pcall(expirationd.task, task_name)
+        if ok then
+            if conf[task_name] then
+                expirationd.task(task_name):stop()
+            else
+                expirationd.task(task_name):kill()
+            end
+        end
+        table.remove(started, i)
+    end
+
+    for task_name, task_conf in pairs(conf) do
+        task_conf = get_task_config(task_conf)
+
+        local skip = task_conf.is_master_only and not opts.is_master
+        if not skip then
+            local task = expirationd.start(task_name, task_conf.space_id,
+                                           task_conf.is_expired,
+                                           task_conf.options)
+            if task == nil then
+                error(role_name .. ": unable to start task " .. task_name)
+            end
+            table.insert(started, task_name)
+        end
+    end
+end
+
+local function stop()
+    for _, task_name in pairs(expirationd.tasks()) do
+        local task = expirationd.task(task_name)
+        task:stop()
+    end
+end
+
+local role = {
+    role_name = role_name,
```

Comment:
Thank you. As far as I understand the only drawback is to use:
`cartridge.service_get('cartridge.roles.expirationd').start()`
instead of just:
`cartridge.service_get('expirationd').start()`

We can take our time. Let's wait for feedback from other users, and then decide what to do.

Reply:
No, cartridge.service_get('expirationd').start() will work anyway because of __index = expirationd - this will lead to search module ref by short name.